File: py/39.py
def solution(a, p):
    try:
        return ((p * p - 2 * p * a) / (2.0 * p - 2 * a)) % 1 == 0
    except ZeroDivisionError:
        return False

def rights(p):
    return len(filter(lambda a: solution(a, p), xrange(3, 1001)))

print max(xrange(120, 1000), key=rights)

File: js/39.js
function rights(p) {
    var r = 0;
    for (var i = 1; i <= 1000; ++i)
        if ((p * p - 2 * p * i) / (2 * p - 2 * i) % 1 == 0)
            ++r;
    return r / 2;
}

var max, mr = 0;

for (var i = 120; i <= 1000; ++i) {
    var r = rights(i);
    if (r > mr)
        mr = r, max = i;
}

console.log(max);
